plugins/woocommerce/src/Blocks/BlockTypes/AddToCartWithOptions.php

```diff
@@ -146,27 +158,20 @@ protected function render( $attributes, $content, $block ) {
 					 * @since 1.5.0
 					 */
 					do_action( 'woocommerce_before_add_to_cart_button' );
-				} elseif ( ProductType::EXTERNAL === $product_type ) {
+				} elseif ( ProductType::VARIABLE === $product_type ) {
 					/**
-					 * Hook: woocommerce_before_add_to_cart_button.
+					 * Hook: woocommerce_before_variations_form.
 					 *
-					 * @since 1.5.0
+					 * @since 2.4.0
 					 */
-					do_action( 'woocommerce_before_add_to_cart_button' );
-				} elseif ( ProductType::GROUPED === $product_type ) {
+					do_action( 'woocommerce_before_variations_form' );
+				} else {
 					/**
 					 * Hook: woocommerce_before_add_to_cart_button.
 					 *
 					 * @since 1.5.0
 					 */
 					do_action( 'woocommerce_before_add_to_cart_button' );
```

Comment:
I wouldn't run this filter for all product types because we can't anticipate if they, in fact, have a product button. If an extension wants this hook or any other to run before/after their template part, they can do it manually (they can use `render_block` or a similar hook). Otherwise, removing the hook for extension that don't want it to run becomes much harder.

Reply:
Ok, that makes sense mostly. I guess I, myself, don't show the add to cart button if my type isn't purchasable. 

I also ran into issues [here](https://github.com/woocommerce/woocommerce/pull/56799/files#diff-cda6779a411f29de14b217e7ac5f8e1ddd839d7600f9d48edcaa6e9202531ad1R237) with core calling functions I didn't want (that's why I wrapped it in that exists check) so don't want to make things hard to remove.

So should I revert this entire chunk?

Reply:
>I think if a custom product type has registered a custom template part, there is no point in running `woocommerce_<product_type>_add_to_cart`, no?

I would agree. I didn't entirely get why you were calling that hook at all. I do use that hook (currently) to register some scripts specific to product types, but for block support I would expect to use new approaches.

Reply:
We are using in WC core product types because some extensions are using them to add extra content, ie when a product is out of stock. But in the case of product types, I think it shouldn't be needed.

Reply:
If I look at my 2 plugins: Mix and Match and Name Your Price... a mix and match container bundle can be name your price enabled... and the nyp price input would appear on the `woocommerce_before_add_to_cart_button` so I would still need a way to enable the compatibility layer for my custom product type. Should I handle that with a compat layer filter on my side?

Reply:
I don't have a strong opinion yet, but I was leaning towards extensions handling the compatibility layer on their own because it gives them more flexibility about when to enable or disable it. For example, for simple products we want to disable it when the product is out of stock or not purchasable (https://github.com/woocommerce/woocommerce/issues/56682). I would expect extensions introducing custom product types might have similar needs, that's why I thought it would make sense to let extensions handle the compatibility layer themselves.

Another option would be to introduce an extra hook that allows extensions to decide whether the compatibility layer is enabled or disabled for certain product types. :thinking: 

What do you think?

Reply:
Makes sense. I would go with this route for now. I think with good documentation it shouldn't be too difficult for extensions to do, and it allows more flexibility. But if we see this being problematic, we can revisit this approach. Does it sound good?